File: src/TellDontAsk.Tests/UseCase/OrderApprovalUseCaseTest.cs
using TellDontAsk.Domain;
using TellDontAsk.Tests.Doubles;
using TellDontAsk.UseCase;
using Xunit;
namespace TellDontAsk.Tests.UseCase
{


    public class OrderApprovalUseCaseTest
    {
        private readonly TestOrderRepository orderRepository;
        private readonly OrderApprovalUseCase useCase;

        public OrderApprovalUseCaseTest()
        {
            orderRepository = new TestOrderRepository();
            useCase = new OrderApprovalUseCase(orderRepository);
        }

        [Fact]
        public void approvedExistingOrder()
        {
            Order initialOrder = new Order();
            initialOrder.setStatus(OrderStatus.CREATED);
            initialOrder.setId(1);
            orderRepository.addOrder(initialOrder);

            OrderApprovalRequest request = new OrderApprovalRequest();
            request.setOrderId(1);
            request.setApproved(true);

            useCase.run(request);

            Order savedOrder = orderRepository.getSavedOrder();
            Assert.Equal(OrderStatus.APPROVED, savedOrder.getStatus());
        }

        [Fact]
        public void rejectedExistingOrder()
        {
            Order initialOrder = new Order();
            initialOrder.setStatus(OrderStatus.CREATED);
            initialOrder.setId(1);
            orderRepository.addOrder(initialOrder);

            OrderApprovalRequest request = new OrderApprovalRequest();
            request.setOrderId(1);
            request.setApproved(false);

            useCase.run(request);

            Order savedOrder = orderRepository.getSavedOrder();
            Assert.Equal(OrderStatus.REJECTED, savedOrder.getStatus());
        }

        [Fact]// (expected = RejectedOrderCannotBeApprovedException.class)
        public void cannotApproveRejectedOrder()
        {
            Order initialOrder = new Order();
            initialOrder.setStatus(OrderStatus.REJECTED);
            initialOrder.setId(1);
            orderRepository.addOrder(initialOrder);

            OrderApprovalRequest request = new OrderApprovalRequest();
            request.setOrderId(1);
            request.setApproved(true);

            Assert.Throws<RejectedOrderCannotBeApprovedException>(() => useCase.run(request));

            Assert.Null(orderRepository.getSavedOrder());
        }

        [Fact]
        public void cannotRejectApprovedOrder()
        {
            Order initialOrder = new Order();
            initialOrder.setStatus(OrderStatus.APPROVED);
            initialOrder.setId(1);
            orderRepository.addOrder(initialOrder);

            OrderApprovalRequest request = new OrderApprovalRequest();
            request.setOrderId(1);
            request.setApproved(false);


            Assert.Throws<ApprovedOrderCannotBeRejectedException>(() => useCase.run(request));

            Assert.Null(orderRepository.getSavedOrder());
        }

        [Fact]
        public void shippedOrdersCannotBeApproved()
        {
            Order initialOrder = new Order();
            initialOrder.setStatus(OrderStatus.SHIPPED);
            initialOrder.setId(1);
            orderRepository.addOrder(initialOrder);

            OrderApprovalRequest request = new OrderApprovalRequest();
            request.setOrderId(1);
            request.setApproved(true);

            Assert.Throws<ShippedOrdersCannotBeChangedException>(() => useCase.run(request));

            Assert.Null(orderRepository.getSavedOrder());
        }

        [Fact]
        public void shippedOrdersCannotBeRejected()
        {
            Order initialOrder = new Order();
            initialOrder.setStatus(OrderStatus.SHIPPED);
            initialOrder.setId(1);
            orderRepository.addOrder(initialOrder);

            OrderApprovalRequest request = new OrderApprovalRequest();
            request.setOrderId(1);
            request.setApproved(false);

            Assert.Throws<ShippedOrdersCannotBeChangedException>(() => useCase.run(request));

            Assert.Null(orderRepository.getSavedOrder());
        }
    }
}

File: php/tests/UseCase/OrderApprovalUseCaseTest.php
<?php

namespace Tests\Pitchart\TellDontAskKata\UseCase;

use PHPUnit\Framework\Assert;
use PHPUnit\Framework\TestCase;
use Pitchart\TellDontAskKata\Domain\Order;
use Pitchart\TellDontAskKata\Domain\OrderStatus;
use Pitchart\TellDontAskKata\UseCase\ApprovedOrderCannotBeRejectedException;
use Pitchart\TellDontAskKata\UseCase\OrderApprovalRequest;
use Pitchart\TellDontAskKata\UseCase\OrderApprovalUseCase;
use Pitchart\TellDontAskKata\UseCase\RejectedOrderCannotBeApprovedException;
use Pitchart\TellDontAskKata\UseCase\ShippedOrdersCannotBeChangedException;
use Tests\Pitchart\TellDontAskKata\Doubles\InMemoryOrderRepository;

class OrderApprovalUseCaseTest extends TestCase
{
    private InMemoryOrderRepository $orderRepository;

    private OrderApprovalUseCase $useCase;

    protected function setUp(): void
    {
        parent::setUp();
        $this->orderRepository = new InMemoryOrderRepository();
        $this->useCase = new OrderApprovalUseCase($this->orderRepository);
    }

    public function test_approve_existing_order()
    {
        $initialOrder = (new Order)->setStatus(OrderStatus::Created)->setId(1);

        $this->orderRepository->addOrder($initialOrder);

        $request = (new OrderApprovalRequest())->setId(1)->setApproved(true);

        $this->useCase->run($request);

        $savedOrder = $this->orderRepository->getSavedOrder();
        Assert::assertEquals(OrderStatus::Approved, $savedOrder->getStatus());
    }


    public function RejectedExistingOrder(): void
    {
        $initialOrder = (new Order())->setStatus(OrderStatus::Created)->setId(1);

        $this->orderRepository->addOrder($initialOrder);

        $request = (new OrderApprovalRequest())->setId(1)->setApproved(false);

        $this->useCase->run($request);

        $savedOrder = $this->orderRepository->getSavedOrder();
        Assert::assertEquals(OrderStatus::Rejected, $savedOrder->getStatus());
    }

    public function test_can_not_approve_rejected_order(): void
    {
        $initialOrder = (new Order())->setStatus(OrderStatus::Rejected)->setId(1);
        $this->orderRepository->addOrder($initialOrder);

        $request = (new OrderApprovalRequest())->setId(1)->setApproved(true);

        $this->expectException(RejectedOrderCannotBeApprovedException::class);

        $this->useCase->run($request);

        Assert::assertNull($this->orderRepository->getSavedOrder());
    }


    public function test_can_not_reject_approved_order(): void
    {
        $initialOrder = (new Order())->setStatus(OrderStatus::Approved)->setId(1);
        $this->orderRepository->addOrder($initialOrder);

        $request = (new OrderApprovalRequest())->setId(1)->setApproved(false);

        $this->expectException(ApprovedOrderCannotBeRejectedException::class);

        $this->useCase->run($request);

        Assert::assertNull($this->orderRepository->getSavedOrder());
    }


    public function test_shipped_orders_cannot_be_rejected(): void
    {
        $initialOrder = (new Order())->setStatus(OrderStatus::Shipped)->setId(1);
        $this->orderRepository->addOrder($initialOrder);

        $request = (new OrderApprovalRequest())->setId(1)->setApproved(false);

        $this->expectException(ShippedOrdersCannotBeChangedException::class);

        $this->useCase->run($request);

        Assert::assertNull($this->orderRepository->getSavedOrder());
    }


}
